### ISSUE TITLE
lua:testmod_lua segfault with clang 3.8

### ISSUE THREAD

**Issue by ingwinlu** [ElektraInitiative/libelektra#1925]
## Steps to Reproduce the Problem

Build on stretch with all bindings, tools, plugins + run tests

## Expected Result

Tests should pass

## Actual Result

testmod_lua segfaults

## System Information

- Elektra Version: master
- Docker Container: #1866 - stretch

## Further Log Files and Output

https://build.libelektra.org/jenkins/blue/organizations/jenkins/elektra-jenkinsfile-scripted/detail/elektra-jenkinsfile-scripted/97/pipeline/85

**Comment by markus2330** [comment]
Thank you for reporting! Is this specific to ASAN?

**Comment by ingwinlu** [comment]
seems like the lua tests run via the lua executable. if the lua swig extension gets build with clang it can not open kdb.so. I will upload my bash history into op.

suggestion: disable swig_lua for clang builds

**Comment by ingwinlu** [comment]
The test is run by ctest via `LD_LIBRARY_PATH=/tmp/workspace/build/lib/  LUA_CPATH=/tmp/workspace/build/src/bindings/swig/lua/?.so /tmp/workspace/build/bin/testmod_lua /tmp/workspace/src/plugins/lua`.

On gcc:
```
jenkins@225c05e4d926:/tmp/workspace/build$ LD_LIBRARY_PATH=/tmp/workspace/build/lib/  LUA_CPATH=/tmp/workspace/build/src/bindings/swig/lua/?.so /tmp/workspace/build/bin/testmod_lua /tmp/workspace/src/plugins/lua
LUA         TESTS
==================

Testing simple variable passing...
[LUA-1] open -->
[LUA-1] get
[LUA-1] <-- close
Testing loading of two active lua plugins...
[LUA-1] open -->
[LUA-2] open -->
[LUA-2] <-- close
[LUA-1] <-- close

========================================================================
NOTE: The following errors are intended. We're testing error conditions!
========================================================================
Testing return values from lua functions...
Testing lua script with syntax error...
There are 1 warnings
buffer is: warnings/#00
number: 11
description: open of plugin returned unsuccessfully (reason contains plugin, see other warnings for details)
ingroup: kdb
module: 
file: /tmp/workspace/src/libs/elektra/plugin.c
line: 294
reason: lua
reason: 
reason: 
number: 131
description: : lua error
ingroup: : plugin
module: : lua
at: /tmp/workspace/src/plugins/lua/lua.cpp:80
reason: : /tmp/workspace/src/plugins/lua/lua/lua_plugin_wrong.lua:2: attempt to call a nil value (global 'wrong')
mountpoint: : 
configfile: : 

test_lua Results: 29 Tests done — 0 errors.
```

on clang:
```
jenkins@055f0905a565:/tmp/workspace/build$ LD_LIBRARY_PATH=/tmp/workspace/build/lib/  LUA_CPATH=/tmp/workspace/build/src/bindings/swig/lua/?.so /tmp/workspace/build/bin/testmod_lua /tmp/workspace/src/plugins/lua
LUA         TESTS
==================

Testing simple variable passing...
[LUA-1] open -->
[LUA-1] get
[LUA-1] <-- close
Testing loading of two active lua plugins...
[LUA-1] open -->
[LUA-2] open -->
[LUA-2] <-- close
[LUA-1] <-- close

========================================================================
NOTE: The following errors are intended. We're testing error conditions!
========================================================================
Testing return values from lua functions...
Testing lua script with syntax error...
There are 1 warnings
buffer is: warnings/#00
number: 11
description: open of plugin returned unsuccessfully (reason contains plugin, see other warnings for details)
ingroup: kdb
module: 
file: /tmp/workspace/src/libs/elektra/plugin.c
line: 294
reason: lua
reason: 
reason: 
number: 131
description: : lua error
ingroup: : plugin
module: : lua
at: /tmp/workspace/src/plugins/lua/lua.cpp:80
reason: : /tmp/workspace/src/plugins/lua/lua/lua_plugin_wrong.lua:2: attempt to call a nil value (global 'wrong')
mountpoint: : 
configfile: : 

test_lua Results: 29 Tests done — 0 errors.
Segmentation fault (core dumped)
```

debugging with gdb: 
```
LUA_CPATH=/tmp/workspace/build/src/bindings/swig/lua/?.so
(gdb) set environment LD_LIBRARY_PATH=/tmp/workspace/build/lib
(gdb) run /tmp/workspace/src/plugins/lua
Starting program: /tmp/workspace/build/bin/testmod_lua /tmp/workspace/src/plugins/lua
warning: Error disabling address space randomization: Operation not permitted
LUA         TESTS
==================

Testing simple variable passing...
[LUA-1] open -->
[LUA-1] get
[LUA-1] <-- close
Testing loading of two active lua plugins...
[LUA-1] open -->
[LUA-2] open -->
[LUA-2] <-- close
[LUA-1] <-- close

========================================================================
NOTE: The following errors are intended. We're testing error conditions!
========================================================================
Testing return values from lua functions...
Testing lua script with syntax error...
There are 1 warnings
buffer is: warnings/#00
number: 11
description: open of plugin returned unsuccessfully (reason contains plugin, see other warnings for details)
ingroup: kdb
module: 
file: /tmp/workspace/src/libs/elektra/plugin.c
line: 294
reason: lua
reason: 
reason: 
number: 131
description: : lua error
ingroup: : plugin
module: : lua
at: /tmp/workspace/src/plugins/lua/lua.cpp:80
reason: : /tmp/workspace/src/plugins/lua/lua/lua_plugin_wrong.lua:2: attempt to call a nil value (global 'wrong')
mountpoint: : 
configfile: : 

test_lua Results: 29 Tests done — 0 errors.
During startup program terminated with signal SIGSEGV, Segmentation fault.
```

so it seems the tests are run, and pass. but there is also a sigsegv from somewhere?

**Comment by ingwinlu** [comment]
all righty: i needed --priviledged on the container to get stacktraces going:

```
(lldb) target create "bin/testmod_lua"
Current executable set to 'bin/testmod_lua' (x86_64).
(lldb) env LUA_CPATH=/tmp/workspace/build/src/bindings/swig/lua/?.so
(lldb) env  LD_LIBRARY_PATH=/tmp/workspace/build/lib/
(lldb) r  /tmp/workspace/src/plugins/lua
Process 24 launched: '/tmp/workspace/build/bin/testmod_lua' (x86_64)
LUA         TESTS
==================

Testing simple variable passing...
[LUA-1] open -->
[LUA-1] get
[LUA-1] <-- close
Testing loading of two active lua plugins...
[LUA-1] open -->
[LUA-2] open -->
[LUA-2] <-- close
[LUA-1] <-- close

========================================================================
NOTE: The following errors are intended. We're testing error conditions!
========================================================================
Testing return values from lua functions...
Testing lua script with syntax error...
There are 1 warnings
buffer is: warnings/#00
number: 11
description: open of plugin returned unsuccessfully (reason contains plugin, see other warnings for details)
ingroup: kdb
module: 
file: /tmp/workspace/src/libs/elektra/plugin.c
line: 294
reason: lua
reason: 
reason: 
number: 131
description: : lua error
ingroup: : plugin
module: : lua
at: /tmp/workspace/src/plugins/lua/lua.cpp:80
reason: : /tmp/workspace/src/plugins/lua/lua/lua_plugin_wrong.lua:2: attempt to call a nil value (global 'wrong')
mountpoint: : 
configfile: : 

test_lua Results: 29 Tests done — 0 errors.
Process 24 stopped
* thread #1: tid = 24, 0x00007ffff62491a0, name = 'testmod_lua', stop reason = signal SIGSEGV: invalid address (fault address: 0x7ffff62491a0)
    frame #0: 0x00007ffff62491a0
error: memory read failed for 0x7ffff6249000
(lldb) bt
* thread #1: tid = 24, 0x00007ffff62491a0, name = 'testmod_lua', stop reason = signal SIGSEGV: invalid address (fault address: 0x7ffff62491a0)
  * frame #0: 0x00007ffff62491a0
    frame #1: 0x0000000000408f74 testmod_lua`llvm_writeout_files + 20
    frame #2: 0x00007ffff6743940 libc.so.6`??? + 240
    frame #3: 0x00007ffff674399a libc.so.6`exit + 26
    frame #4: 0x00007ffff672e2e8 libc.so.6`__libc_start_main + 248
    frame #5: 0x0000000000403cfa testmod_lua`_start + 42
(lldb) 

```

**Comment by ingwinlu** [comment]
and one more:
```
* thread #1: tid = 31, 0x000000000040498e testmod_lua`main(argc=<unavailable>, argv=<unavailable>) + 3006 at testmod_lua.c:137, name = 'testmod_lua', stop reason = step over
    frame #0: 0x000000000040498e testmod_lua`main(argc=<unavailable>, argv=<unavailable>) + 3006 at testmod_lua.c:137 [opt]
   134 		test_fail ();
   135 		test_wrong ();
   136 	
-> 137 		print_result ("test_lua");
   138 	
   139 		return nbError;
   140 	}
(lldb) n

test_lua Results: 29 Tests done — 0 errors.
Process 31 stopped
* thread #1: tid = 31, 0x00000000004049c4 testmod_lua`main(argc=<unavailable>, argv=<unavailable>) + 3060 at testmod_lua.c:139, name = 'testmod_lua', stop reason = step over
    frame #0: 0x00000000004049c4 testmod_lua`main(argc=<unavailable>, argv=<unavailable>) + 3060 at testmod_lua.c:139 [opt]
   136 	
   137 		print_result ("test_lua");
   138 	
-> 139 		return nbError;
   140 	}
(lldb) n
Process 31 stopped
* thread #1: tid = 31, 0x00007ffff672e2e1 libc.so.6`__libc_start_main + 241, name = 'testmod_lua', stop reason = step over
    frame #0: 0x00007ffff672e2e1 libc.so.6`__libc_start_main + 241
libc.so.6`__libc_start_main:
->  0x7ffff672e2e1 <+241>: movl   %eax, %edi
    0x7ffff672e2e3 <+243>: callq  0x7ffff6743980            ; exit
    0x7ffff672e2e8 <+248>: xorl   %eax, %eax
    0x7ffff672e2ea <+250>: jmp    0x7ffff672e227            ; <+55>
(lldb) n
Process 31 stopped
* thread #1: tid = 31, 0x00007ffff672e2e3 libc.so.6`__libc_start_main + 243, name = 'testmod_lua', stop reason = instruction step over
    frame #0: 0x00007ffff672e2e3 libc.so.6`__libc_start_main + 243
libc.so.6`__libc_start_main:
->  0x7ffff672e2e3 <+243>: callq  0x7ffff6743980            ; exit
    0x7ffff672e2e8 <+248>: xorl   %eax, %eax
    0x7ffff672e2ea <+250>: jmp    0x7ffff672e227            ; <+55>
    0x7ffff672e2ef <+255>: movq   0x37e3da(%rip), %rax
(lldb) n
Process 31 stopped
* thread #1: tid = 31, 0x00007ffff62491a0, name = 'testmod_lua', stop reason = signal SIGSEGV: invalid address (fault address: 0x7ffff62491a0)
    frame #0: 0x00007ffff62491a0
error: memory read failed for 0x7ffff6249000
```

looks like the return is failing for clang... Pretty much at the end of my knowledge here. Any Ideas?

**Comment by markus2330** [comment]
It could be undefined behavior in print_result but it is much more likely that the lua interpreter leaves something in an unexpected state. Does the same error happen with every lua interpreter? If any other lua interpreter works, let us simply use that one.

**Comment by ingwinlu** [comment]
no dice with either lua 5.3, 5.2 or 5.1.

I also could not pinpoint the problem to a specific part of the test. If i commented out parts of it the error changed but I never got it to run through. Actually the most successful is the current state because it actually only crashes at the return...

could any of the memory leaks reported in the asan testcase for the testcase have something to do with it? Most seem unrelated to the specific test or even elektra though...

This issue is kind of blocking #1866 hence my interest in solving it.

**Comment by ingwinlu** [comment]
oh and @markus2330 print_result passes. the return segfaults.

**Comment by markus2330** [comment]
> no dice with either lua 5.3, 5.2 or 5.1.

and luajit?

> could any of the memory leaks reported in the asan testcase for the testcase have something to do with it? Most seem unrelated to the specific test or even elektra though...

Memory leaks cannot (afaik) cause such a crash, undefined behavior, however, may cause stack corruption which could cause something like this. (Usually gdb detects stack corruption so maybe something more complicated is involved.)

But trying to run it with valgrind may help you because it also reports other errors next to memory leaks.

> This issue is kind of blocking #1866 hence my interest in solving it.

Please do not get blocked by such low priority bugs. Simply disable the lua plugin for clang 3.8.

> print_result passes. the return segfaults.

This does not mean much. The corruption of the stack and the crash are usually not the same place.

But as said, it is much more likely lua does something weird than print_result.

**Comment by ingwinlu** [comment]
I opened up a PR illustrating the build issue
https://build.libelektra.org/jenkins/blue/organizations/jenkins/elektra-jenkinsfile/detail/elektra-jenkinsfile/202/pipeline/267

**Comment by ingwinlu** [comment]
Resolved, we no longer run the test in master.
